Improve sentences
----------------------------

In this notebook we want to improve the corpora of new sentences and the final corpora which is the concatenation of the original corpora and the new one (we got the two in the following notebook [new_sentences](extract_new_sentences.ipynb)). 

For the improvement we want to make the following modifications that we found to be necessary after fine-tuning the `t5` model on the new sentences:

- Remove the definitions: We had already extract the new sentences without the definitions we will only need to load the "true sentences".
- Replace exclamation marks (!) at the end of the phrases by period marks (.): This will help the transformers to identify the end of a sentence. Not that we will do this for both of the new sentences and the last version corpora (v4). We cannot know if it is the right thing to do in order to obtain better performance that's why we will create a function and add it when making hyper-parameter search and reduce the search space if necessary.
- Remove sentences with less than 3 tokens in one of the French of Wolof version. This will reduce the number of sentences but can provide better agreement with the right way to write sentences.

The above modifications on the version 4 of the corpora and the diagne's sentences will be necessary to create the version 5 of the corpora. The additional version of the diagne's sentences and 6 of the corpora will understand additionally sentences got from `omniglot`. We will not select the sentences by length for the new versions.

All the modifications can be done with the `pandas` library.

In [1]:
import re
import pandas as pd
from tokenizers.pre_tokenizers import Whitespace
from wolof_translate.utils.database_manager import TranslationMongoDBManager

### Remove the definitions

Let us load the two corpora.

In [2]:
corpora_v4 = pd.read_csv('data/extractions/new_data/corpora_v4.csv')

diagne_sentences = pd.read_csv('data/extractions/new_data/sentences.csv')

In [3]:
corpora_v4.shape[0], diagne_sentences.shape[0]

(2500, 1615)

### Add periods to the end of the sentences

We have three possibilities:

- We can replace the exclamations by period marks (option 2)
- ... add period mark at each end of sentences (option 3)
- ... do the two previous modification at same time (option 4)

The following function will be added inside the `TransformerSequences` class when recuperating the sentences before the training.

In [4]:
%%writefile wolof-translate/wolof_translate/utils/improvements/end_marks.py

import pandas as pd
from typing import *

def add_end_mark(sentences: Union[list, str], end_mark: str = ".", end_mark_to_remove: Union[str, None] = None, 
                 replace: bool = False, poss_end_marks: list = ['!', '?', '.', '...']):
    
    if isinstance(sentences, str): sentences = [sentences]
    
    if replace and end_mark_to_remove is None:
        
        raise ValueError("You must provide a end mark to remove if you want to make replacement !")
    
    new_sentences = []
    
    # if replace is chosen we will replace the end to remove by the end mark if not only add end mark to each sentence and remove the end mark to remove
    for sentence in sentences:
        
        if replace:
            
            if sentence[-1] == end_mark_to_remove:
                
                sentence = sentence[:-1].strip() + end_mark
            
        else:
            
            if not end_mark_to_remove is None and sentence[-1] == end_mark_to_remove:
                
                sentence = sentence[:-1]
            
            if not sentence[-1] in poss_end_marks: 
                
                sentence = sentence.strip() + end_mark
        
        new_sentences.append(sentence)
    
    return new_sentences
    

Overwriting wolof-translate/wolof_translate/utils/improvements/end_marks.py


Let us try the different type of end mark adding with the above function.

In [5]:
%run wolof-translate/wolof_translate/utils/improvements/end_marks.py

In [6]:
# recuperate some sentences from the diagne's sentences
sample = diagne_sentences.sample(30)

In [7]:
sample

french  \
1583  Les enfants veulent venir et les adultes souha...   
949                            J'ai vu Monsieur Moussa.   
251                    Tout village Sérère est propre !   
612                                  Vous êtes ceux-là.   
868                                       Que veux-tu ?   
125                        Aucune dame ne s'est égarée.   
835              Donnes le livre à ce fils bien élevé !   
584                           Qui d'autre veut partir ?   
1516         C'est Fatim, si on va au fond des choses !   
892                    Tu n'avais pas été à l'intérieur   
77                      Qu'il attrappe quelles vaches ?   
4                             Homme n'est pas mauvais !   
244                       J'ai vu la maison de son ami.   
428                                    Quand il partira   
133                           Cherche un autre cheval !   
122                          Aucun lion ne s'est égaré.   
1160                        Chez les Sérère et les Peul   
18                              Tu vois, cet homme là ?   
320                           Cherche-les à l'intérieur   
801                  C'est qu'il déteste l'impolitesse.   
62                    Quelle jeune fille s'est égarée ?   
745                               Ngor peut-il partir ?   
364                                 L'homme qui eut été   
654                                     Ainsi comment ?   
1136               Je croie qu'aujourd'hui il viendra !   
322                                              Sors !   
1453               Je ne te répéterai plus jamais rien.   
658                                        Quel autre ?   
1329                            J'ai vu vache d'Europe.   
748                            Qui t'a donné ce livre ?   

                                                  wolof  
1583  Xale yi bëgg nañu dikk, te mag ni ñaan nañu ŋg...  
949                              Gis naa góor gi Musaa.  
251                        Béppu dëkku Séeréer set na !  
612                                        Ñooñu ŋgeen.  
868                                          Loo bëgg ?  
125                                 Jenn jigéen réerul.  
835            Joxal téere bi doomu nit ku yaru kooku !  
584                           Keneen kan moo bëgg dem ?  
1516                             Soo demee, Faatim la !  
892                                 Demuloowoon ca biir  
77                              Mu japp nag yooyu yan ?  
4                                           Nit bonul !  
244                                Gis naa kër xaritam.  
428                                          Su dee dem  
133                                 Seetal fas weneen !  
122                                 Gayndé genn réerul.  
1160                            Ci Séeréer yi ag Pël yi  
18                                 Gis ŋga nit kookee ?  
320                                    Wut leen ci biir  
801                                   Dafa bañ rewande.  
62                                    Ndaw san a réer ?  
745                                          Ŋgor dem ?  
364                                     Góor gi demkoon  
654                                        Noonee nan ?  
1136                            Defe naa tày dana ñëw !  
322                                            Génnal !  
1453                            Waxati ma la koon dara.  
658                                        Leneen lan ?  
1329                                Gis naa nagu Tugël.  
748                               Ku la may bii téere ?

We need to add the modification on each version of the sentences.

- Replace exclamations by end marks.

In [8]:
french_sents = sample['french'].to_list()

wolof_sents = sample['wolof'].to_list()

french_sents = add_end_mark(french_sents, end_mark_to_remove='!', replace=True)

wolof_sents = add_end_mark(wolof_sents, end_mark_to_remove='!', replace=True)

pd.DataFrame({'french': french_sents, 'wolof': wolof_sents})


french  \
0   Les enfants veulent venir et les adultes souha...   
1                            J'ai vu Monsieur Moussa.   
2                     Tout village Sérère est propre.   
3                                  Vous êtes ceux-là.   
4                                       Que veux-tu ?   
5                        Aucune dame ne s'est égarée.   
6               Donnes le livre à ce fils bien élevé.   
7                           Qui d'autre veut partir ?   
8           C'est Fatim, si on va au fond des choses.   
9                    Tu n'avais pas été à l'intérieur   
10                    Qu'il attrappe quelles vaches ?   
11                           Homme n'est pas mauvais.   
12                      J'ai vu la maison de son ami.   
13                                   Quand il partira   
14                           Cherche un autre cheval.   
15                         Aucun lion ne s'est égaré.   
16                        Chez les Sérère et les Peul   
17                            Tu vois, cet homme là ?   
18                          Cherche-les à l'intérieur   
19                 C'est qu'il déteste l'impolitesse.   
20                  Quelle jeune fille s'est égarée ?   
21                              Ngor peut-il partir ?   
22                                L'homme qui eut été   
23                                    Ainsi comment ?   
24                Je croie qu'aujourd'hui il viendra.   
25                                              Sors.   
26               Je ne te répéterai plus jamais rien.   
27                                       Quel autre ?   
28                            J'ai vu vache d'Europe.   
29                           Qui t'a donné ce livre ?   

                                                wolof  
0   Xale yi bëgg nañu dikk, te mag ni ñaan nañu ŋg...  
1                              Gis naa góor gi Musaa.  
2                         Béppu dëkku Séeréer set na.  
3                                        Ñooñu ŋgeen.  
4                                          Loo bëgg ?  
5                                 Jenn jigéen réerul.  
6             Joxal téere bi doomu nit ku yaru kooku.  
7                           Keneen kan moo bëgg dem ?  
8                               Soo demee, Faatim la.  
9                                 Demuloowoon ca biir  
10                            Mu japp nag yooyu yan ?  
11                                         Nit bonul.  
12                               Gis naa kër xaritam.  
13                                         Su dee dem  
14                                 Seetal fas weneen.  
15                                Gayndé genn réerul.  
16                            Ci Séeréer yi ag Pël yi  
17                               Gis ŋga nit kookee ?  
18                                   Wut leen ci biir  
19                                  Dafa bañ rewande.  
20                                  Ndaw san a réer ?  
21                                         Ŋgor dem ?  
22                                    Góor gi demkoon  
23                                       Noonee nan ?  
24                             Defe naa tày dana ñëw.  
25                                            Génnal.  
26                            Waxati ma la koon dara.  
27                                       Leneen lan ?  
28                                Gis naa nagu Tugël.  
29                              Ku la may bii téere ?

- Add end marks at each sentence without replacing or removing

In [9]:
french_sents = sample['french'].to_list()

wolof_sents = sample['wolof'].to_list()

french_sents = add_end_mark(french_sents)

wolof_sents = add_end_mark(wolof_sents)

pd.DataFrame({'french': french_sents, 'wolof': wolof_sents})


french  \
0   Les enfants veulent venir et les adultes souha...   
1                            J'ai vu Monsieur Moussa.   
2                    Tout village Sérère est propre !   
3                                  Vous êtes ceux-là.   
4                                       Que veux-tu ?   
5                        Aucune dame ne s'est égarée.   
6              Donnes le livre à ce fils bien élevé !   
7                           Qui d'autre veut partir ?   
8          C'est Fatim, si on va au fond des choses !   
9                   Tu n'avais pas été à l'intérieur.   
10                    Qu'il attrappe quelles vaches ?   
11                          Homme n'est pas mauvais !   
12                      J'ai vu la maison de son ami.   
13                                  Quand il partira.   
14                          Cherche un autre cheval !   
15                         Aucun lion ne s'est égaré.   
16                       Chez les Sérère et les Peul.   
17                            Tu vois, cet homme là ?   
18                         Cherche-les à l'intérieur.   
19                 C'est qu'il déteste l'impolitesse.   
20                  Quelle jeune fille s'est égarée ?   
21                              Ngor peut-il partir ?   
22                               L'homme qui eut été.   
23                                    Ainsi comment ?   
24               Je croie qu'aujourd'hui il viendra !   
25                                             Sors !   
26               Je ne te répéterai plus jamais rien.   
27                                       Quel autre ?   
28                            J'ai vu vache d'Europe.   
29                           Qui t'a donné ce livre ?   

                                                wolof  
0   Xale yi bëgg nañu dikk, te mag ni ñaan nañu ŋg...  
1                              Gis naa góor gi Musaa.  
2                        Béppu dëkku Séeréer set na !  
3                                        Ñooñu ŋgeen.  
4                                          Loo bëgg ?  
5                                 Jenn jigéen réerul.  
6            Joxal téere bi doomu nit ku yaru kooku !  
7                           Keneen kan moo bëgg dem ?  
8                              Soo demee, Faatim la !  
9                                Demuloowoon ca biir.  
10                            Mu japp nag yooyu yan ?  
11                                        Nit bonul !  
12                               Gis naa kër xaritam.  
13                                        Su dee dem.  
14                                Seetal fas weneen !  
15                                Gayndé genn réerul.  
16                           Ci Séeréer yi ag Pël yi.  
17                               Gis ŋga nit kookee ?  
18                                  Wut leen ci biir.  
19                                  Dafa bañ rewande.  
20                                  Ndaw san a réer ?  
21                                         Ŋgor dem ?  
22                                   Góor gi demkoon.  
23                                       Noonee nan ?  
24                            Defe naa tày dana ñëw !  
25                                           Génnal !  
26                            Waxati ma la koon dara.  
27                                       Leneen lan ?  
28                                Gis naa nagu Tugël.  
29                              Ku la may bii téere ?

- Remove exclamation marks and add period marks

In [10]:
french_sents = sample['french'].to_list()

wolof_sents = sample['wolof'].to_list()

french_sents = add_end_mark(french_sents, end_mark_to_remove='!')

wolof_sents = add_end_mark(wolof_sents, end_mark_to_remove='!')

pd.DataFrame({'french': french_sents, 'wolof': wolof_sents})


french  \
0   Les enfants veulent venir et les adultes souha...   
1                            J'ai vu Monsieur Moussa.   
2                     Tout village Sérère est propre.   
3                                  Vous êtes ceux-là.   
4                                       Que veux-tu ?   
5                        Aucune dame ne s'est égarée.   
6               Donnes le livre à ce fils bien élevé.   
7                           Qui d'autre veut partir ?   
8           C'est Fatim, si on va au fond des choses.   
9                   Tu n'avais pas été à l'intérieur.   
10                    Qu'il attrappe quelles vaches ?   
11                           Homme n'est pas mauvais.   
12                      J'ai vu la maison de son ami.   
13                                  Quand il partira.   
14                           Cherche un autre cheval.   
15                         Aucun lion ne s'est égaré.   
16                       Chez les Sérère et les Peul.   
17                            Tu vois, cet homme là ?   
18                         Cherche-les à l'intérieur.   
19                 C'est qu'il déteste l'impolitesse.   
20                  Quelle jeune fille s'est égarée ?   
21                              Ngor peut-il partir ?   
22                               L'homme qui eut été.   
23                                    Ainsi comment ?   
24                Je croie qu'aujourd'hui il viendra.   
25                                              Sors.   
26               Je ne te répéterai plus jamais rien.   
27                                       Quel autre ?   
28                            J'ai vu vache d'Europe.   
29                           Qui t'a donné ce livre ?   

                                                wolof  
0   Xale yi bëgg nañu dikk, te mag ni ñaan nañu ŋg...  
1                              Gis naa góor gi Musaa.  
2                         Béppu dëkku Séeréer set na.  
3                                        Ñooñu ŋgeen.  
4                                          Loo bëgg ?  
5                                 Jenn jigéen réerul.  
6             Joxal téere bi doomu nit ku yaru kooku.  
7                           Keneen kan moo bëgg dem ?  
8                               Soo demee, Faatim la.  
9                                Demuloowoon ca biir.  
10                            Mu japp nag yooyu yan ?  
11                                         Nit bonul.  
12                               Gis naa kër xaritam.  
13                                        Su dee dem.  
14                                 Seetal fas weneen.  
15                                Gayndé genn réerul.  
16                           Ci Séeréer yi ag Pël yi.  
17                               Gis ŋga nit kookee ?  
18                                  Wut leen ci biir.  
19                                  Dafa bañ rewande.  
20                                  Ndaw san a réer ?  
21                                         Ŋgor dem ?  
22                                   Góor gi demkoon.  
23                                       Noonee nan ?  
24                             Defe naa tày dana ñëw.  
25                                            Génnal.  
26                            Waxati ma la koon dara.  
27                                       Leneen lan ?  
28                                Gis naa nagu Tugël.  
29                              Ku la may bii téere ?

### Remove sentences with less than 3 tokens

Only the words interest us and we want to separate them by space or marks. Notice also that we want verify for each version and if one it is not accepted then the another one also.

In [11]:
def select_by_length(sentences: pd.DataFrame, min_len: int = 4):
    
    new_sentences = {'french': [], 'wolof': []}
    
    for i in list(sentences.index):
        
        sent1 = sentences.loc[i, 'french']
        
        sent2 = sentences.loc[i, 'wolof']
        
        tokens1 = re.findall('\w+', sent1)
        
        tokens2 = re.findall('\w+', sent2)
        
        take = True
        
        if len(tokens1) < min_len or len(tokens2) < min_len:
            
            take = False
        
        if take:
            
            new_sentences['french'].append(sent1)
            
            new_sentences['wolof'].append(sent2)
            
    return pd.DataFrame(new_sentences)

Let us make an example with the sample that we got earlier.

In [12]:
sample = select_by_length(sample, min_len=3)


In [13]:
sample

french  \
0   Les enfants veulent venir et les adultes souha...   
1                            J'ai vu Monsieur Moussa.   
2                    Tout village Sérère est propre !   
3                        Aucune dame ne s'est égarée.   
4              Donnes le livre à ce fils bien élevé !   
5                           Qui d'autre veut partir ?   
6          C'est Fatim, si on va au fond des choses !   
7                    Tu n'avais pas été à l'intérieur   
8                     Qu'il attrappe quelles vaches ?   
9                       J'ai vu la maison de son ami.   
10                                   Quand il partira   
11                          Cherche un autre cheval !   
12                         Aucun lion ne s'est égaré.   
13                        Chez les Sérère et les Peul   
14                            Tu vois, cet homme là ?   
15                          Cherche-les à l'intérieur   
16                 C'est qu'il déteste l'impolitesse.   
17                  Quelle jeune fille s'est égarée ?   
18                                L'homme qui eut été   
19               Je croie qu'aujourd'hui il viendra !   
20               Je ne te répéterai plus jamais rien.   
21                            J'ai vu vache d'Europe.   
22                           Qui t'a donné ce livre ?   

                                                wolof  
0   Xale yi bëgg nañu dikk, te mag ni ñaan nañu ŋg...  
1                              Gis naa góor gi Musaa.  
2                        Béppu dëkku Séeréer set na !  
3                                 Jenn jigéen réerul.  
4            Joxal téere bi doomu nit ku yaru kooku !  
5                           Keneen kan moo bëgg dem ?  
6                              Soo demee, Faatim la !  
7                                 Demuloowoon ca biir  
8                             Mu japp nag yooyu yan ?  
9                                Gis naa kër xaritam.  
10                                         Su dee dem  
11                                Seetal fas weneen !  
12                                Gayndé genn réerul.  
13                            Ci Séeréer yi ag Pël yi  
14                               Gis ŋga nit kookee ?  
15                                   Wut leen ci biir  
16                                  Dafa bañ rewande.  
17                                  Ndaw san a réer ?  
18                                    Góor gi demkoon  
19                            Defe naa tày dana ñëw !  
20                            Waxati ma la koon dara.  
21                                Gis naa nagu Tugël.  
22                              Ku la may bii téere ?

It only kept 26 samples. Let us make the same for all the sentences.

In [14]:
diagne_sentences = select_by_length(diagne_sentences, min_len=3)

corpora_v4 = select_by_length(corpora_v4, min_len=3)

Let us verify how many sentences we have kept.

In [15]:
# for diagne's sentences
diagne_sentences.shape[0]

1328

In [16]:
# in corpora v4
corpora_v4.shape[0]

2212

We removed $651$ sentences from the corpora v4 and $645$ sentences from the diagne's sentences.

Let us save the results to make again hyperparameter search with them.

In [17]:
# save the new version of the diagne's sentences in red_sentences.csv
diagne_sentences.to_csv('data/extractions/new_data/red_sentences.csv', index = False)

# save the new version of the corpora
corpora_v4.to_csv('data/extractions/new_data/corpora_v5.csv', index = False)

### Recuperate additional sentences 

In [18]:
corpora_v4 = pd.read_csv('data/extractions/new_data/corpora_v4.csv')

diagne_sentences = pd.read_csv('data/extractions/new_data/sentences.csv')

Let us recuperate additional sentences to augment the corpora.

- First file:

In [19]:
path = 'data/additional_documents/omniglot/sentences_french_wolof.txt'

with open(path, 'r', encoding='utf-8') as f:
    
    # recuperate lines
    omn_1 = f.readlines()
    
# recuperate sentences
omn_1_ = {
    'french': [],
    'wolof': []
}

for omn in omn_1:
    
    omn = omn.split('\t')
    
    sentences = [s.strip() for s in omn if s.strip() != '']
    
    omn_1_['french'].append(sentences[0])
    
    omn_1_['wolof'].append(sentences[1])

In [20]:
pd.options.display.max_rows = 74
omn_1_ = pd.DataFrame(omn_1_).head(74)

# print the first sentences from omniglot
omn_1_.head(74)

french  \
0                                    Bienvenue   
1                                        Salut   
2                                        Salut   
3                                        Salut   
4                             Comment vas-tu ?   
5                              Es-tu en paix ?   
6                             Comment vas-tu ?   
7                     Paix seulement, et toi ?   
8      Je suis seulement toi, comment vas-tu ?   
9      Ça fait longtemps que je ne t'ai pas vu   
10                     Comment t'appelles-tu ?   
11                     Comment t'appelles-tu ?   
12                     Comment t'appelles-tu ?   
13                             Mon nom est ...   
14                             Mon nom est ...   
15                           Je m'appelles ...   
16                             D'où viens-tu ?   
17                           D'où venez-vous ?   
18                             Je viens de ...   
19                             Je viens de ...   
20                    Plaisir de te rencontrer   
21                    Content de te rencontrer   
22            Je suis content de te rencontrer   
23                                   Bon matin   
24                            Bonne après-midi   
25                                Bonne soirée   
26                                   Au revoir   
27                                   Au revoir   
28                              Bonne chance !   
29                               Bonne santé !   
30                      Aies une bonne journée   
31                                  Bon voyage   
32                              Comprends-tu ?   
33                                Je comprends   
34                         Je ne comprends pas   
35                        Je ne le connais pas   
36                           Je ne le sais pas   
37      S'il te plaît, parles plus lentement !   
38                     Répètes s'il te plaît !   
39                       Ecris s'il te plaît !   
40                         Parles-tu anglais ?   
41                           Parles-tu Wolof ?   
42                                 Oui, un peu   
43            Comment dis t-on ceci en Wolof ?   
44                                Excuse-moi !   
45                          Combien ça coute ?   
46                                Pardonne-moi   
47                                  Si tu veux   
48                                       Merci   
49                                       Viens   
50                               Venez en paix   
51                     Où sont les toilettes ?   
52  Ce garçon peut te payer pour n'import quoi   
53             Cette jeune fille va tout payer   
54             Accepte-tu de danser avec moi ?   
55             Accepte-tu de danser avec moi ?   
56                               Tu me manques   
57                                  Je t'aimes   
58                                Je t'admires   
59                                  Je t'aimes   
60            Je te souhaites de récupérer tôt   
61                                  Vas t'en !   
62                                  Vas t'en !   
63                    Laisses moi tranquille !   
64                    Laisses moi tranquille !   
65                                  À l'aide !   
66                                       Feu !   
67                                    Arrête !   
68                         Appelle la police !   
69                           Mes voeux de noël   
70                      Mes voeux de nouvel an   
71                         Mes voeux de paques   
72                        Mes voeux de tabaski   
73                    Mes voeux d'anniversaire   

                                          wolof  
0                                        Merhbe  
1                                    Na nga def  
2                                  Na ngeen def  
3                                Salaam aleekum  
4                                 Jaam 

- Second file

In [21]:
path = 'data/additional_documents/omniglot/sentences_french_wolof_2.txt'

with open(path, 'r', encoding='utf-8') as f:
    
    # recuperate lines
    omn_2 = f.readlines()
    
# recuperate sentences
omn_2_ = {
    'french': [],
    'wolof': []
}

for omn in omn_2:
    
    omn = omn.split('–')
    
    sentences = [s.strip() for s in omn if s.strip() != '']
    
    omn_2_['wolof'].append(sentences[0])
    
    omn_2_['french'].append(sentences[1])

In [22]:
pd.options.display.max_rows = 74
omn_2_ = pd.DataFrame(omn_2_).head(74)

# print the first sentences from omniglot
omn_2_.head(74)

french                    wolof
0          Comment allez-vous ?           Na ngeen def ?
1                Bonne matinée.         Jaam nga fanane.
2                   Bonne nuit.           Fanaanal jaam.
3             À une autre fois.               Ba beneen.
4                 Je t'en prie.                   Agsil.
5              Je vous en prie.        Agsileen ak jaam.
6                       Pardon.                 Baal ma.
7                          Oui.                     Wau.
8                          Non.                 Deh-det.
9               Es-tu en paix ?            Jaam nga am ?
10             Comment vas-tu ?            Jaam nga am ?
11              Paix seulement.                Jaam rek.
12             Je suis en paix.                Jaam rek.
13                     Et toi ?                Yow nag ?
14      C'est comment ton nom ?          Naka-nga sant ?
15            D'où venez-vous ?        Fan ngeen joghe ?
16            Je viens des USA.   Maa ngi joghe les USA.
17           Parles-tu Anglais?         Deg nga Angale ?
18         Parles-tu Français ?        Deg nga Faranse ?
19  Je parle seulement anglais.    Angale rekk laa degg.
20     Je parle un peu anglais.  Degg naa tuuti Faranse.
21      Je ne parles pas Wolof.     Mahn deggumah Wolof.
22   Je ne parles pas Français.   Mahn deggumah Faranse.
23            Je n'entends pas.                 Degguma.
24               J'aimerais … .        Dama bahggoon … .
25                   Où est … ?              Fahn la … ?
26                Est-ce loin ?               Soreh na ?
27            C'est devant toi.               Cha kanam.
28                  Devant toi.               Cha kanam.
29                      Gauche.                Chammoon.
30                       Droit.                Ndeyjoor.
31                 Mais entre !           Dugghal waay !
32            Combine ça vaut ?               Lii naata?
33             C'est trop cher.           Seer na torob.

- Third file

In [23]:
path = 'data/additional_documents/omniglot/sentences_french_wolof_3.txt'

with open(path, 'r', encoding='utf-8') as f:
    
    # recuperate lines
    omn_3 = f.readlines()
    
# recuperate sentences
omn_3_ = {
    'french': [],
    'wolof': []
}

for omn in omn_3:
    
    omn = omn.split('\t')
    
    sentences = [s.strip() for s in omn if s.strip() != '']
    
    omn_3_['wolof'].append(sentences[0])
    
    omn_3_['french'].append(sentences[1])
      
        

In [24]:
pd.options.display.max_rows = 74
omn_3_ = pd.DataFrame(omn_3_).head(74)

# print the first sentences from omniglot
omn_3_.head(74)

french                             wolof
0             Que la paix soit avec toi.                 Salaam maaleekum.
1       Que la paix soit avec toi aussi.                  Maaleekum salaam
2                        Es-tu en paix ?                     Jaama ngaam ?
3                        Paix seulement.                        Jamma rek.
4                Comment va ta famille ?                  Ana waa kër gi ?
5    Où sont les membres de ta famille ?                  Ana waa kër gi ?
6                       Ils sont là-bas.                         Ñunga fa.
7             J'espère qu'ils vont bien.                 Mbaa defuñu dara.
8               Rien ne va mal avec eux.                 Mbaa defuñu dara.
9                    Non, ils vont bien.             Déedéet, defuñu dara.
10              Rien ne va mal chez-eux.             Déedéet, defuñu dara.
11         Comment se passe le travail ?                  Naka ligéey bi ?
12             J'y suis avec, pas à pas.  Maangi ci kowam, ndanka, ndanka.
13         J'y suis avec, petit à petit.  Maangi ci kowam, ndanka, ndanka.
14                      Comment vas-tu ?                       Nanga def ?
15                      Comment vas-tu ?                         Noo def ?
16  Comment vont les gens de la maison ?                 Naka waa kër ga ?
17               Comment va la famille ?                 Naka njaboot gi ?
18                         Je vais bien.                    Maangi fi rek.
19        Qu'est-ce que tu as rapporté ?                     Lan nga indi?
20          J'ai rapporté des nouvelles.                  Xibaar laa indi.
21               Je suis dans la maison.                 Maangi ci nég bi.
22               Ils se tiennent dehors.              Ñungi taxaw ci biti.
23         As-tu passé la nuit en paix ?                Jaama nga fanaan ?
24         Comment se passe la matinée ?                   Naka subba si ?
25             La matinée se passe bien.               Subba sangi fi rek.
26      As-tu passé la matinée en paix ?                   Jaama nga endu?
27       Comment se passe l'après-midi ?                   Naka bëcëk bi ?
28           L'après-midi se passe bien.              Bëcëk baangi fi rek.
29         Comment se passes la soirée ?                   Naka ngoon si ?
30              La soirée se passe bien.              Ngoon saangi fi rek.
31            Comment se passe la nuit ?                   Naka guddi gi ?
32                La nuit se passe bien.               Guddi gaangi fi rek
33                          Comment ...?                         Noo ... ?

- Let us retrieve sentences, that was created by internet users in our following platform [addition_sentences_platform](https://french-wolof-translation.streamlit.app/), from our MongoDB Atlas Database.

In [25]:
# initialize the path to the csv files
sentences_path = 'data/additional_documents/hugging_face/wolof_french.csv'
deleted_path = 'data/additional_documents/hugging_face/deleted.csv'

# initialize the server uri and database name
uri = "mongodb+srv://oumar199:Jacksparrow360@woloftranslationcluster.u0gk7.mongodb.net/?retryWrites=true&w=majority"
# uri = "mongodb+srv://oumarkanetest:Jacksparrow360__@woloftranslationcluster.u0gk7.mongodb.net/?retryWrites=true&w=majority"
database_name = 'WolofTranslation'

# initialize the database manager 
database_manager = TranslationMongoDBManager(uri, database_name)


# save locally the data sets
database_manager.save_data_frames(sentences_path, deleted_path)

hug_sents = pd.read_csv(sentences_path)

In [26]:
# # print some documents inside the sentences collection using the uri
# def write_documents(uri):
        
#     from pymongo import MongoClient
    
#     # connect to the database
#     client = MongoClient(uri)
    
#     # get the database
#     db = client['WolofTranslation']
    
#     # get the sentences collection
#     collection = db['sentences']
    
#     # recuperate some documents
#     documents = collection.find()
    
#     # write the documents in a file
#     for doc in documents:
        
#         print(str(doc))
#         print(len(documents))
            
#     # close the connection
#     client.close()
    
# write_documents(uri)


In [27]:
pd.options.display.max_rows = 74

hug_sents.head(74)

french  \
0   J'en suis sûr, cette photo ci c'est la photo p...   
1                    J'arrive tout de suite chez toi.   
2   Je vois devant moi une photo sur laquelle beau...   
3   Ceux-ci sont des personnes qui sont sortis pou...   
4   Sur la photo, ont voit des personnes qui se ré...   
5   On voit sur la photo beaucoup de personnes sor...   
6   Je pense que la photo que je vois là, c'est un...   
7   Bonjour ! Ceci est une photo. Sur cette photo ...   
8   Ceci est une théière. Une théière dans la quel...   
9   Ce que j'ai en face de moi, sur la photo, c'es...   
10  Sur la photo que j'ai, j'ai vu une sorte de ch...   
11  Sur cette photo je suis en face de quelque cho...   
12  Ceci est une photo sur laquelle je vois un poi...   
13  Sur cette photo ci, j'y ai vu des chambres fai...   
14  Cette photo ci montre une maison qui contient ...   
15  Oui ça, il semble que ce soit une fleur, mais ...   
16  À l'intérieur de cette photo je vois beaucoup ...   
17  Je vois une large chambre avec un lit pour pas...   
18  Sur la photo qu'on m'a montrée, c'est une cham...   
19  Ceci ressemble à un bahut. Un bahut chocolat d...   
20  Sur la photo que vous m'avez envoyée il y a de...   
21  Je sens que ceux-là sont, vraisemblablement, d...   
22  Ceci est ce qu'on appelle une douche, qu'on a ...   
23  À l'intérieur de la photo, j'ai vu trois beaux...   
24    J'ai vu quelque chose qu'on achète au marché...   
25  Sur la photo que vous m'avez envoyée, j'y ai v...   
26  Sur cette photo ci, il y a des pierres qui son...   
27  C'est des enfants qui jouent avec leurs jouets...   
28  Ce que tu m'as envoyé jusqu'à maintenant, la p...   
29  J'ai vu ici une photo sur laquelle il y a un l...   
30  C'est cela qu'on appelle un berceau. Il est co...   
31  C'est l'entrée de Louga. C'est à la frontière ...   
32  Sur cette photo il semble qu'il y est un table...   
33  Ceux-là sont des garçons montés sur des chevau...   
34  Celui-là que tu vois c'est le dirigeant du pay...   
35  Sur cette photo, il y a un gendarme monté sur ...   
36  Sur cette photo il y a le Capitaine Seydina Om...   
37  Sur cette photo, c'est une personne de couleur...   
38  Jusqu'à présent, on voit sur la photo des sold...   
39  Sur cette photo ci, j'y vois des gendarmes por...   
40  J'ai vu ici des gendarmes qui portent leurs te...   
41  Ça c'est un terrain de jeu, c'est-à-dire un tr...   
42  Cette photo ci est une photo sur laquelle il y...   
43  Sur cette photo, il y a une chose qui ressembl...   
44  Ceci, cependant, est un bol. Il y a du sucre d...   
45  Ceci est une maison. La maison est, cependant,...   
46  Sur cette photo, il y a d'une salle dans la qu...   
47  Sur cette photo ci que je vois il y a des pers...   
48  Sur la photo que j'ai vue, j'y ai vu une jolie...   
49  Ceci on dirait que c'est une théière, une théi...   
50  Ceci est une théière ; une théière sur la quel...   
51  Ceci est un ancien milieu avec une constructio...   
52  Sur cette photo on voit ce qu'on appelle 'pois...   
53  Sur la photo que j'ai eue a voir il y a assuré...   
54  Oui mon parent, je vois que ça c'est des pots ...   
55  Sur la photo que vous m'avez envoyée, j'ai vu,...   
56  Sur cette photo il y a des gens qui sont assis...   
57  Ceux-ci sont de vieilles cases à l'intérieur d...   
58  Oui, ceci est une chambre. Une large chambre d...   
59  Sur cette photo, c'est <mask> dans laquelle on...   
60  Sur cette photo ci il y a une porte avec des c...   
61  Ceci est un four. C'est ce qu'on appelle une c...   
62  Sur cette photo ci, j'y aperçois du beurre ; i...   
63  Ceci est une bouche qu'on a conçue, dans laque...   
64  Ici, cependant, il semble qu'il y est une port...   
65  Sur la photo qu'on m'a envoyée, j'y ai vu un c...   
66  Sur ma photo qu'on m'a envoyée, il y a apparem...   
67  Sur cette photo, je suis face à une chaise pei...   
68  Ceux-là, cependant, sont des cases. Celle qui ...   
69  Ceux-là sont, semble-t-il, des paniers de cour...   


Let us add them into the diagne's sentences and the version 4 of the corpora and remove duplicated rows.

- On corpora version 4

In [28]:
corpora_v4 = pd.concat((corpora_v4, omn_1_, omn_2_, omn_3_, hug_sents), axis=0)

# drop duplicated rows
print(sum(corpora_v4.duplicated()))

7


In [29]:
corpora_v4.drop_duplicates(inplace=True)

7 rows were deleted.

- On corpora version 5

In [30]:
diagne_sentences = pd.concat((diagne_sentences, omn_1_, omn_2_, omn_3_, hug_sents), axis=0)

# drop duplicated rows
print(sum(diagne_sentences.duplicated()))

4


In [31]:
diagne_sentences.drop_duplicates(inplace=True)

4 rows were deleted.

### Web Scraping

This step involve to scrape some websites to obtain more sentences using `selenium` and `BeautifulSoup`. It will require to meticulously investigate each website and identify the elements to trigger and to extract. We will also need to identify the transformation to add to the different sentences. All the transformations, excepted the sentences' capitalization, are already implemented and can be directly used. 

We will scrape the following websites: 

- url1: [corporan.huma-num.fr](https://corporan.huma-num.fr/Lexiques/indexM.php?controle=afficheDico&action=main&ficXML=DicoWolof.xml&langue=Wolof)

In [54]:
url1 = "https://corporan.huma-num.fr/Lexiques/indexM.php?controle=afficheDico&action=main&ficXML=DicoWolof.xml&langue=Wolof"

#### Corporan.huma-num scraping

Let us add the necessary libraries.

In [32]:
from wolof_translate.utils.extract_new_sentences import NewSentenceExtraction
from wolof_translate.utils.sent_transformers import TransformerSequences
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait
from wolof_translate.utils.sent_corrections import *
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome import service
from selenium.webdriver.common.by import By
from urllib import request as rq
from selenium import webdriver
from bs4 import BeautifulSoup
from tqdm import tqdm
from typing import *
import pandas as pd

We need to identify the element containing the sentences. It is easy in that case since they are identified by their classes:

- The french sentences are identified by the `xe` class
- The wolof sentences by the `xv` class.

Since the website consists of a dictionary of word, we must click on one of them to display the sentences. The words' elements are not links but it use a ajax function. Then we must use trigger event to display the different pages. The words are represented by the `div` tag and each of them has a unique id which is simply the number of the word. We have `9965` words so the ids range from 1 to 9965.

Let us begin by initializing the drive. We will use the chrome driver in our case.

In [2]:
# initialize the option
options = webdriver.ChromeOptions()

# we will add the headless option which will not render the browser
options.add_argument('--headless')

# let us recuperate the service
s = service.Service('C:\Program Files\Chromedriver\chromedriver.exe')

# let us recuperate the driver
driver = webdriver.Chrome(service=s, options=options)



Let us request to go the website.

In [57]:
# send a get request to the website
driver.get(url1)

Let us initialize the classes.

In [33]:
# initialize the class map
class_map = {
    'xe': 'french',
    'xv': 'wolof'
}


Let us initialize the transformer.

In [34]:
transformer = TransformerSequences(remove_mark_space, delete_guillemet_space, add_mark_space)

Let us create a function which get recuperate the content and with beautiful soup recuperate the sentences. We will use an implicit wait at each time since we don't know what element to wait for: Not all pages contain sentences. We will wait 0.2 seconds before getting the sentences.

In [60]:
def get_sentences_from_page(driver: webdriver.chrome.webdriver.WebDriver, sentences: Dict[str, list], transformer: Union[TransformerSequences, None] = None, replaces: dict = {'…': '...'}, end_marks = ['.', '!', '?', '...']):
    
    # let us recuperate the content of the current page
    content = driver.page_source
    
    # recuperate the document object
    doc = BeautifulSoup(content, 'lxml')
    
    # recuperate the elements
    elements = {class__: doc.find_all('span', class_ = class__) for class__ in sentences}
    
    if elements != {key: [] for key in sentences}:
        
        # count the number of sentences that we got
        sent_count = len(elements[list(sentences)[0]])
        
        # for each group of sentences we will add some transformations
        for s in range(sent_count):
            
            # initialize the common end mark
            end_mark = None
            
            # recuperate the sentences and add to the dictionary of sentences
            for class__ in sentences:
                
                # recuperate the current sentence
                sentence = elements[class__][s].text.strip()
                
                # replace marks
                for replace in replaces:
                    
                    sentence = sentence.replace(replace, replaces[replace])
                
                if not sentence[-1] in end_marks:
                    
                    # add an end mark if the end mark is got from one of the previous sentences
                    if not end_mark is None:
                        
                        sentence += end_mark
                    
                    else: # add a period mark
                        
                        sentence += '.'
                
                else: # if an end mark is identified recuperated it as the end mark of the other sentences
                    
                    end_mark = sentence[-1]
                
                # capitalize the current sentence
                sentence = sentence[0].upper() + sentence[1:]
                
                # add transformations to the sentence
                if transformer: sentence = transformer(sentence)[0]
                
                # add the sentence to the dictionary of sentences
                sentences[class__].append(sentence)
            
    # return the sentences
    return sentences
    

In [ ]:
# calculate the approximate number of minutes
0.5*9956/60

82.96666666666667

Initialize the sentences 🛑.

In [35]:
# recuperate the saved sentences
scraped_sentences = pd.read_csv('data/extractions/new_data/scraped_sents.csv').to_dict('list')

# initialize the sentences
sentences = {key: scraped_sentences[class_map[key]] for key in class_map}

# current_id
current_id = 9836

# initialize the ids
word_ids = list(range(current_id, 9966))

Let us recuperate the sentences by clicking recursively at each word.

In [ ]:
# initialize the time sleep previously introduced
time_sleep = 0.6

# initialize a save step
save_step = 20

# recuperate the sentences of each word
for id in tqdm(word_ids):
        
    # click next to the word
    word = driver.find_element(By.ID, id)
    driver.execute_script('arguments[0].click();', word)
    
    # add a time sleep
    driver.implicitly_wait(time_sleep)
    
    # recuperate the sentences
    sentences = get_sentences_from_page(driver, sentences, transformer, replaces={'…': '...', '\xa0': ' ', 'qqn': "quelqu'un"})
    
    # save the sentences
    if id % save_step == 0:
        
        pd.DataFrame({value: sentences[key] for key, value in class_map.items()}).to_csv('data/extractions/new_data/scraped_sents.csv', index = False)
    
    

100%|██████████| 130/130 [03:41<00:00,  1.70s/it]


Not all sentences were recuperated. Let us make it again to recuperate additional sentences from the website.

Initialize the sentences 🛑.

In [68]:
# recuperate the saved sentences
scraped_sentences_2 = pd.read_csv('data/extractions/new_data/scraped_sents2.csv').to_dict('list')

# initialize the sentences
sentences_2 = {key: scraped_sentences_2[class_map[key]] for key in class_map}

# current_id
current_id = 300

# initialize the ids
word_ids = list(range(current_id, 9966))

Let us recuperate the sentences by clicking recursively at each word.

In [66]:
# initialize the time sleep previously introduced
time_sleep = 0.6

# initialize a save step
save_step = 20

# recuperate the sentences of each word
for id in tqdm(word_ids):
        
    # click next to the word
    word = driver.find_element(By.ID, id)
    driver.execute_script('arguments[0].click();', word)
    
    # add a time sleep
    driver.implicitly_wait(time_sleep)
    
    # recuperate the sentences
    sentences_2 = get_sentences_from_page(driver, sentences_2, transformer, replaces={'…': '...', '\xa0': ' ', 'qqn': "quelqu'un"})
    
    # save the sentences
    if id % save_step == 0:
        
        pd.DataFrame({value: sentences_2[key] for key, value in class_map.items()}).to_csv('data/extractions/new_data/scraped_sents2.csv', index = False)
    
    

100%|██████████| 9666/9666 [5:19:39<00:00,  1.98s/it]   


Let us take concatenate the two data frames and drop duplicated rows.

In [70]:
# concatenate the translations
scraped_sentences = pd.concat((pd.DataFrame(scraped_sentences), pd.DataFrame(scraped_sentences_2)))

In [72]:
# drop duplicated translations
scraped_sentences.drop_duplicates(inplace=True)

Let us verify how many translations we have.

In [74]:
# print the number of new translations
scraped_sentences.shape[0]

9388

Let us save the result.

In [76]:
scraped_sentences.to_csv('data/extractions/new_data/scraped_sents.csv', index=False)

We obtained `9388` translations from the first scraped website. Let us now make pre-processing on it. We will based on the following strings to make it:
- The parentheses containing 'ou...' 
- The slashes '/'
- The parentheses: We mostly want to delete them all
- Replace 'qqch' by 'quelque chose'
- Replace ';.' by '.'
- Replace 'œ' by 'oe'
- Identify the sentence with only ')' but not '('.

Let us create a function which can recuperate the sentences with regular expressions and another function which can replace an expression by another.

In [45]:
# %%writefile wolof-translate/wolof_translate/utils/re_extract.py
from typing import *
import re

def extract_with_re(translations: pd.DataFrame, columns: list = [], regex: Union[str, None] = None, recup_csv: str = 're_extract.csv'):
    
    # get all columns if no column is given
    if not columns: columns = list(translations.columns)
    
    for column in columns:
        
        assert column in set(translations.columns)
    
    # if not regular expression is given then return in the csv file all the sentences
    if not regex is None:
        
        # initialize the dictionary of the concerned sentences
        re_sents = {
            column: [] for column in columns
        }
        
        # at each index we will verify if a column respect the regular expression
        for i in range(translations.shape[0]):
            
            # make verification for each column
            re_true = False
            
            for column in columns:
                
                if re.match(regex, translations.loc[i, column]):
                    
                    re_true = True
                
            # add the line inside the set of translations to investigate
            if re_true:
                
                for column in columns:
                        
                        re_sents[column].append(translations.loc[i, column])
                
                # remove the sentences from the DataFrame
                translations.drop(index=[i], inplace=True)
            
        # save the resulted sentences in order to modify (it will wait until you finish the modifications)
        re_sents = pd.DataFrame(re_sents)
        
        print("You recuperated the following sentences:")
        print(re_sents)
        
        re_sents.to_csv(recup_csv, index=False)
        
        alert_ = ''
        
        while True:
            
            finish = input(f"{alert_}Provide Yes(y) if you finish and No(n) if not yet!")
            
            if finish == 'y':
                
                break
            
            elif finish != 'n':
                
                alert_ = "You cannot provide a command which is different from y (for Yes) and n (No)"
            
        # recuperate the modifications
        re_sents_2 = pd.read_csv(recup_csv)
        
        # add the sentences inside the original data frame and return the results
        translations = pd.concat((re_sents_2, translations))
        
        translations.index = range(translations.shape[0])
        
        return translations, re_sents_2
    
    return None
    
def replace_expression(translations: pd.DataFrame, columns: list = [], regex: Union[str, None] = None, new_expression: str = ''):
    
    # get all columns if no column is given
    if not columns: columns = list(translations.columns)
    
    for column in columns:
        
        assert column in set(translations.columns)
    
    # if not regular expression is given then return in the csv file all the sentences
    if not regex is None:
        
        # initialize the dictionary of the concerned sentences
        re_sents = {
            column: [] for column in columns
        }
        
        # at each index we will verify if a column respect the regular expression before making replacement
        for i in range(translations.shape[0]):
            
            for column in columns:
                
                expressions = re.findall(regex, translations.loc[i, column])
                
                if expressions != []:
                
                    translations.loc[i, column] = re.sub(regex, new_expression, translations.loc[i, column]).strip()

                    for column in columns:
                        
                        re_sents[column].append(translations.loc[i, column])
            
        # save the resulted sentences in order to modify (it will wait until you finish the modifications)
        re_sents = pd.DataFrame(re_sents)
        
        print("You recuperated the following sentences:")
        print(re_sents)
    
        return translations, re_sents
    
    return None

- Let us recuperate all translations containing parentheses with 'ou' and verify which of them will need to separated as multiple translations.

In [21]:
scraped_sentences, re_sents = extract_with_re(pd.DataFrame(scraped_sentences), regex='.*\(ou.*\).*')

You recuperated the following sentences:
                                               french                                      wolof
0   Je m'occupe de toi; je suis à toi. (ou encore)...                            Maa ngi ci yow.
1               C'est bien (ou bien) d'accord; ça va.                                   Baax na.
2   Il est bel et bien au courant (ou bien) il l'a...                    Yég na ko ba bëgga dee.
3                C'est pareil (ou bien) c'est unique.                                   Benn la.
4    Ils sont semblables (ou bien) ils ne font qu'un.                                 Benn lañu.
..                                                ...                                        ...
27         Être soupe au lait (ou bien) être frustré.                                  Tàng xol.
28                                       De bon cœur.  Ci xol bu sedd (ou bien) ci xol bu tàlli.
29       Vous êtes égaux (ou bien) vous êtes pareils.                                 

- Let us recuperate all translations containing slashes.

In [34]:
scraped_sentences, re_sents = extract_with_re(scraped_sentences, regex='.*(/).*')


You recuperated the following sentences:
                                              french                                              wolof
0  Qu'est-ce qui te prend ? / qu'est-ce qui t'arr...                                 (loc.) Lu la dal ?
1  Fais des boulettes plus grosses / fais davanta...                                    Yokkal dank yi.
2    Oh la, Omar ! / Arrête, Omar ! (Tu as fini ?) .                                         Eey, Omar.
3  J'en ai plus que toi / j'en ai un plus grand (...                                      Maa la ëpple.
4                Le voilà ! / Voilà / nous-y voilà !                                  (Lov.)Mu ngoogu !
5  Celui qui meurt au marché fait lui-même son an...  Ku dee marse, yaa tàgge sa bopp // ku dee ca j...
6  Il n'a pas mangé / il n'est pas en train de ma...                                            Lekkul.
7  Je n'ai pas mangé / je ne suis pas en train de...                                 Lekkuma lekkul ma.


- Separate in multiple ones the translations with parentheses

In [37]:
scraped_sentences, re_sents = extract_with_re(scraped_sentences, regex='.*\(.*\).*')

You recuperated the following sentences:
                                                french                                              wolof
0        Ces touffes (de cheveux), c'est leur coutume.                            Jubb yii, seen aada la.
1    Avant de tirer sur une bête, je vois d'abord s...  Bala may fital aw rab, dinaa xool ndax mu ngi ...
2       L'interdiction est temporaire (ne durera pas).                                   Aaye bi du yàgg.
3    Il prête sa charrette (à qui veut) uniquement ...  Li muy abalaate saretam bi yépp dey, bëgga fal...
4                      Les enfants sont arrivés (ici).                                 Xale yi agsi nañu.
..                                                 ...                                                ...
167  J'ai appelé ce village Njar-Meew (lait coupé) ...  Lu naqari la mu ma doon def, di ma ko yéjje, m...
168  Il dit qu'on a libéré les invités (ils peuvent...                           Mu ne yewwi nañu gan yi.
169  

- Delete parentheses

In [38]:
scraped_sentences, re_sents = replace_expression(scraped_sentences, regex='\(.*\)')

You recuperated the following sentences:
                                                french                                              wolof
0                    Ces touffes , c'est leur coutume.                            Jubb yii, seen aada la.
1    Avant de tirer sur une bête, je vois d'abord s...  Bala may fital aw rab, dinaa xool ndax mu ngi ...
2                      L'interdiction est temporaire .                                   Aaye bi du yàgg.
3     Il prête sa charrette  uniquement pour être élu.  Li muy abalaate saretam bi yépp dey, bëgga fal...
4                           Les enfants sont arrivés .                                 Xale yi agsi nañu.
..                                                 ...                                                ...
168                 J'ai appelé ce village Njar-Meew .  Lu naqari la mu ma doon def, di ma ko yéjje, m...
169                Il dit qu'on a libéré les invités .                           Mu ne yewwi nañu gan yi.
170  

- Replace 'qqch' by 'quelque chose'

In [47]:
scraped_sentences, re_sents = replace_expression(scraped_sentences, regex='qqch', new_expression='quelque chose')


You recuperated the following sentences:
                                               french                                              wolof
0   Il y a quelque chose  dans mes habits; tiens l...  Am na luy dox ci sama yére yi; téyeel ma xale ...
1   Dis-moi quelque chose qui puisse apporter la p...         Wax ma lu mëna amal jàmm ci sunu réew mi !
2   Je pensais qu'il m'appelait pour quelque chose...   Dama defe woon ne lu am-maanaa la ma doon wooye.
3   Le réparateur est arrivé. Qui aurait quelque c...        Baay-jagal dikk na. Ku amoon looy defarlu ?
4   Ses propos ont été brefs; lui aurais-tu fait q...       Waxam ji dafa bar; xanaa danga ko def dara ?
..                                                ...                                                ...
42  Quand on fait quelque chose, il est normal d'e...                            Kuy xalam, di ci jaayu.
43  Quand on reçoit quelque chose d'une personne q...  Bu la ku xér di may, sa xol sedd; waaye bu nge...
44  Ce dont on

- Replace ';.' by '.'

In [49]:
scraped_sentences, re_sents = replace_expression(scraped_sentences, regex=';\.', new_expression='.')

You recuperated the following sentences:
                                                french                                              wolof
0    Il y a quelque chose  dans mes habits.tiens l'...  Am na luy dox ci sama yére yi; téyeel ma xale ...
1    Il y a quelque chose  dans mes habits.tiens l'...  Am na luy dox ci sama yére yi.téyeel ma xale b...
2    Ses propos ont été brefs.lui aurais-tu fait qu...       Waxam ji dafa bar; xanaa danga ko def dara ?
3    Ses propos ont été brefs.lui aurais-tu fait qu...        Waxam ji dafa bar.xanaa danga ko def dara ?
4    Ça, il faudra tout faire pour l'avoir, c'est q...  Lii dey, nanga ci gar sa bakkan.lu lay amal nj...
..                                                 ...                                                ...
443           Il est fâché.il ne veut même pas manger.                       Dafa aru; nanguwul sax lekk.
444           Il est fâché.il ne veut même pas manger.                        Dafa aru.nanguwul sax lekk.
445  

In [53]:
# function created to add correction 
# def correction(text: str):
    
#     charac = ''
    
#     for i in range(len(text[:-1])):
        
#         letter = text[i]
        
#         if letter == '.' and re.match('[a-zA-Z]', text[i+1]):
            
#             charac = charac + '; '
        
#         else:
            
#             charac = charac + letter
        
#     return charac + text[-1]

- Replace 'œ' by 'oe'

In [58]:
scraped_sentences, re_sents = replace_expression(scraped_sentences, regex='œ', new_expression='oe')


You recuperated the following sentences:
                                               french                                       wolof
0   Une femme de moeurs légères ne se marie pas fa...                 Jigéen juy fo du yomba séy.
1                            Elle est ma jeune soeur.                               Sama rakk la.
2                                       De bon coeur.                            Ci xol bu tàlli.
3                                       De bon coeur.                             Ci xol bu sedd.
4                      Chante, nous ferons le choeur.                           Woyal, nu awu la.
..                                                ...                                         ...
93              Ton souhait ne te vient pas du coeur.                Sa yéene ji àggul ci sa xol.
94                           Coupe le jusqu'au coeur.                   Gor ko ba àgg ca yeññ wa.
95  Va jeter un coup d'oeil pour voir s'il est rév...            Demal yër ba

- Identify translations with only a closing parentheses

In [59]:
scraped_sentences, re_sents = extract_with_re(scraped_sentences, regex='.*\).*')


You recuperated the following sentences:
                                               french                                              wolof
0   La marmite, si elle doit être délicieuse, quan...                Cin, su naree neex, bu baxee, xeeñ.
1        L'homme et ses désirs, Dieu et sa décision).              Nit ak bëgg-bëggam, Yàlla ak dogalam.
2   C'est avec l'esprit avec lequel tu as creusé l...     Xel mi nga gase teen bi la ci keneen di naane.
3   Une langue ne mange pas son semblable, si on n...           Làmmiñ du lekk morom ma, te toggeesu ko.
4   Si ton talent de cavalier te pousse à monter s...  Su la mën ngawar jayee ba nga mafñàndum saaw, ...
..                                                ...                                                ...
26                      Ton corps est-il en paix ?) .                            Sa yaram jàmm bonjour !
27  Généreux, avec tes biens) sois généreux, mais ...                               Yéwén : ci sa alal !
28  Qui a broy

Let us save the sentences and delete the space between a point or a comma and the word behind them. 

In [60]:
scraped_sentences.to_csv('data/extractions/new_data/scraped_sents.csv', index=False)

Let us add the scraped translations inside the diagne's translations and the corpora (version 4).

In [39]:
shape_1 = diagne_sentences.shape[0]

In [40]:
# inside diagne's sentences
diagne_sentences = pd.concat((diagne_sentences, pd.DataFrame(scraped_sentences)))

# inside the corpora
corpora_v4 = pd.concat((corpora_v4, pd.DataFrame(scraped_sentences)))

# inside diagne's sentences (we want 3000 translations for diagne's sentences)
# diagne_sentences = pd.concat((diagne_sentences, pd.DataFrame(scraped_sentences).sample(3000 - shape_1)))


In [41]:
diagne_sentences.index = range(diagne_sentences.shape[0])
corpora_v4.index = range(corpora_v4.shape[0])

Let us add the scraped sentences inside a new collection and add it also the translations from the already existing collection.

In [13]:
scraped_sentences = pd.DataFrame(scraped_sentences)

database_manager.insert_documents([{
        '_id': id,
        'french': scraped_sentences.loc[id, 'french'],
        'wolof': scraped_sentences.loc[id, 'wolof'],
    } for id in scraped_sentences.index], 'sentences_plus')

Let us insert inside the collection the translations from the sentences collection.

In [16]:
hug_sents.index = range(scraped_sentences.shape[0], scraped_sentences.shape[0] + hug_sents.shape[0])
database_manager.insert_documents([{
        '_id': id,
        'french': hug_sents.loc[id, 'french'],
        'wolof': hug_sents.loc[id, 'wolof'],
    } for id in hug_sents.index], 'sentences_plus')

---------------

Let us consider only the fourth end mark adding's method and verify if we have duplicated rows.

- Corpora v4

In [36]:
french_sents = corpora_v4['french'].to_list()

wolof_sents = corpora_v4['wolof'].to_list()

french_sents = add_end_mark(french_sents, end_mark_to_remove='!')

wolof_sents = add_end_mark(wolof_sents, end_mark_to_remove='!')

corpora_v4_end = pd.DataFrame({'french': french_sents, 'wolof': wolof_sents})

In [37]:
duplicated = corpora_v4_end.duplicated()

sum(duplicated)

5

We obtain 5 duplicated rows. Let us recuperate their indexes.

In [38]:
dup_indexes = duplicated[duplicated == True].index

In [39]:
dup_indexes

Int64Index([1428, 1597, 2241, 2243, 2367], dtype='int64')

Let us remove those indexes from the corpora.

In [40]:
corpora_v4.drop(index = dup_indexes, inplace=True)

- Diagne's sentences

In [41]:
french_sents = diagne_sentences['french'].to_list()

wolof_sents = diagne_sentences['wolof'].to_list()

french_sents = add_end_mark(french_sents, end_mark_to_remove='!')

wolof_sents = add_end_mark(wolof_sents, end_mark_to_remove='!')

diagne_end = pd.DataFrame({'french': french_sents, 'wolof': wolof_sents})

In [42]:
duplicated = diagne_end.duplicated()

sum(duplicated)

5

We obtain 5 duplicated rows. Let us recuperate their indexes.

In [43]:
dup_indexes = duplicated[duplicated == True].index

In [44]:
dup_indexes

Int64Index([546, 715, 1359, 1361, 1485], dtype='int64')

Let us remove those indexes from the corpora.

In [45]:
diagne_sentences.drop(index = dup_indexes, inplace=True)

Apply transformations on all sentences.

In [46]:
# apply the transformations
def correction(text: str):
    
    text = transformer(text)[0]
    
    return text

# add corrections to the translations
diagne_sentences['french'] = diagne_sentences['french'].apply(correction)
diagne_sentences['wolof'] = diagne_sentences['wolof'].apply(correction)

corpora_v4['french'] = corpora_v4['french'].apply(correction)
corpora_v4['wolof'] = corpora_v4['wolof'].apply(correction)

# print clearly the results
diagne_sentences.head(74)


french  \
0                        Il n'a pas encore été   
1                   Il n'a pas encore été voir   
2           Il ne s'est pas encore substitué à   
3           Il n'a pas encore pris de contacts   
4                    Homme n'est pas mauvais !   
5                  L'homme n'est pas mauvais !   
6                        Quelqu'un est parti !   
7                            Homme est parti !   
8                         Un homme est parti !   
9                                J'ai vu lion.   
10                            J'ai vu un lion.   
11                             C'est un homme.   
12                      Cet homme près de moi.   
13                   Cet homme là près de toi.   
14                  Tu vois cet homme là-bas ?   
15                        Tu sais, cet homme ?   
16                      Tu vois cet homme là ?   
17                     Tu sais, cet homme là ?   
18                     Tu vois, cet homme là ?   
19                            J'ai vu l'homme.   
20                           J'ai vu la femme.   
21                         Regarde ce lutteur.   
22             Regarde ce semblant de lutteur.   
23                       Regarde cette aumône.   
24                          Tu as vu l'homme ?   
25             Tu as vu, auparavant, l'homme ?   
26             Tu connaissais la jeune fille ?   
27              Tu reconnais cet individu-là ?   
28                 Tu avais reconnu la femme ?   
29                          Cet homme a étudié   
30                 Ces jeunes gens sont sages.   
31          Je parle de cette dame que voilà !   
32                       Je parle de ce petit.   
33            Cette femme-ci n'est pas partie.   
34    Ces enfants que voilà ne sont pas sages.   
35              Tu connais cet enfant là-bas ?   
36                          Cette dame là-bas.   
37  C'est ces livres là-bas que j'ai achetés !   
38       Cause avec cet homme qui est à côté !   
39                     Tu vois ces femmes-là ?   
40                    Nourris cette vache-ci !   
41                   Tu connais cet homme-là ?   
42                  Tu as vu ces jeunes gens ?   
43    Je parle de la personne qui est sortie !   
44                   Tu avais vu cette femme ?   
45            Je connaissais cet individu-là !   
46             J'avais reconnu cette dame-là !   
47                           J'ai vu l'enfant.   
48                   Je parle de cet homme-ci.   
49          Ce n'est pas là une belle parole !   
50                Cet homme là-bas que tu vois   
51       C'est ces livres-là qu'il m'a donnés.   
52                          J'ai vu un mouton.   
53                       J'ai aperçu un homme.   
54                       J'ai aperçu un homme.   
55                       J'ai aperçu des gens.   
56                    Quel homme s'est égaré ?   
57                   Quel mouton s'est égaré ?   
58                   Quel cheval s'est égaré ?   
59                 Quelle femme s'est égarée ?   
60                   Quelle clé s'est égarée ?   
61                 Quelle chose s'est égarée ?   
62           Quelle jeune fille s'est égarée ?   
63         Quelles personnes se sont égarées ?   
64              Quels moutons se sont égarés ?   
65              Quels chevaux se sont égarés ?   
66            Quelles femmes se sont égarées ?   
67              Quelles clés se sont égarées ?   
68            Quelles choses se sont égarées ?   
69             Quels couteaux se sont égarés ?   
70                  Tu parles de quel mouton ?   
71                    Tu parles de ce mouton ?   
72                               Quel mouton ?   
73            Quels moutons avais-tu achetés ?   

                                  wolof  
0                               Demagul  
1                             Seeteegul  
2                             Wuutóogul  
3                              Giséegul  
4                           Nit bonul !  
5                           Nit bonul

In [47]:
corpora_v4.head(74)

french  \
0   Tout être humain est le résultat d'un père et ...   
1   On peut ne pas les reconnaître, ne pas les aim...   
2   Mais ils sont là, avec leur visage, leurs atti...   
3   J'ai longtemps rêvé que ma mère était noire. J...   
4   Puis j'ai découvert, lorsque mon père, à l'âge...   
5                    Cela a été difficile à admettre.   
6   Il m'a fallu retourner en arrière, recommencer...   
7     En souvenir de cela, j'ai écrit ce petit livre.   
8   De ce visage que j'ai reçu à ma naissance, j'a...   
9                D'abord, qu'il m'a fallu l'accepter.   
10  Affirmer que je ne l'aimais pas serait lui don...   
11  Je ne le haïssais pas, je l'ignorais, je l'évi...   
12  Pendant des années, je crois que je ne l'ai ja...   
13  Sur les photos, je détournais les yeux, comme ...   
14  À l'âge de huit ans à peu près, j'ai vécu en A...   
15  Dans la case que nous habitions (le mot case a...   
16  De ce temps, pour ainsi dire consécutivement, ...   
17  Mon corps, le corps de ma mère, le corps de mo...   
18  Leur stature, leurs seins lourds, la peau luis...   
19    Le sexe des garçons, leur gland rose circoncis.   
20  Des visages sans doute, mais comme des masques...   
21  Les ventres saillants, le bouton du nombril pa...   
22  L'odeur des corps aussi, le toucher, la peau n...   
23  J'ai cette impression de la grande proximité, ...   
24  En Afrique, l'impudeur des corps était magnifi...   
25  Elle donnait du champ, de la profondeur, elle ...   
26  Elle s'harmonisait avec le pays Ibo, avec le t...   
27  Elle brillait dans ces noms qui entraient en m...   
28  Elle imprégnait la muraille de la forêt pluvie...   
29  Quand on est enfant, on n'use pas de mots (et ...   
30  Je suis en ce temps-là très loin des adjectifs...   
31  Je ne peux pas dire ni même penser : admirable...   
32              Mais je suis capable de le ressentir.   
33  À quel point les arbres aux troncs rectilignes...   
34  L'entrée dans Obudu, je m'en souviens bien : l...   
35  Mon père a arrêté son auto, avec ma mère il do...   
36  Je suis seul au milieu de la foule, je n'ai pa...   
37  Les mains me touchent, passent sur mes bras, s...   
38  Parmi tous ceux qui se pressent autour de moi,...   
39  Quand ma mère revient (peut-être vaguement inq...   
40  Comment aurais-je pu imaginer que cette femme ...   
41  Je me rappelle seulement cette question : « Es...   
42  Et non la réponse – sans doute rassurante, peu...   
43  L'Afrique, c'était le corps plutôt que le visage.   
44  L'Afrique, c'était la violence des sensations,...   
45  Le premier souvenir que j'ai de ce continent, ...   
46  Je suis dans la cabine du bateau qui longe len...   
47  L'Afrique qui déjà m'ôtait mon visage me renda...   
48  Ce que je recevais dans le bateau qui m'entraî...   
49  La guerre, le confinement dans l'appartement d...   
50  Désormais, pour moi, il y aurait avant et aprè...   
51     La liberté à Ogoja, c'était le règne du corps.   
52  Illimité, le regard, du haut de la plate-forme...   
53  Si je fais un effort de mémoire, je puis recon...   
54  Quelqu'un qui aurait gardé la mémoire photogra...   
55                              Sans doute un jardin.   
56  Non pas un jardin d'agrément – existait-il dan...   
57  Plutôt un espace d'utilité, où mon père avait ...   
58  Quelque part vers l'arrière de la maison, au m...   
59  Un jardin, soit, puisqu'un des employés de la ...   
60  À l'autre bout du terrain, il devait y avoir l...   
61  Dans cette cour, il y avait surtout des enfant...   
62  Tout cela pourrait donner l'impression d'une v...   
63  Pourtant c'était la liberté totale du corps et...   
64  Devant la maison, dans la direction opposée à ...   
65  Au Sud, la pente conduisait à un affluent de l...   
66  Vers le Nord et l'est, je pouvais voir la gran...   
67  Je me souviens de la violence en Afrique. Non ...   
68  Cette violence-là n'était pas vraiment physiqu...   
69  J'en avais le corps miné, des quintes de toux ...   


Let us save the results.

In [48]:
# save the results
diagne_sentences.to_csv('data/extractions/new_data/ad_sentences.csv', index=False)
corpora_v4.to_csv('data/extractions/new_data/corpora_v6.csv', index=False)

Let us verify how many sentences we got for the moment.

In [49]:
# from diagne's book
diagne_sentences.shape[0]

1979

In [50]:
# inside the corpora
corpora_v4.shape[0]

2861

We recuperated `11435` sentences from the diagne's book and the final corpora contains `12317` sentences.

The training can take a long time to finish. We can take only a sample in place of all the translations. Let us take `5000` samples to begin and a seed of `0` for the generator.

In [51]:
# # recuperate the sentences
# diagne_sentences = pd.read_csv('data/extractions/new_data/ad_sentences.csv')
# corpora_v4 = pd.read_csv('data/extractions/new_data/corpora_v6.csv')

In [112]:
diagne_sentences_s = diagne_sentences.sample(5000)
corpora_v4_s = corpora_v4.sample(5000)

Let us save again the translations.

In [ ]:
diagne_sentences_s.to_csv('data/extractions/new_data/ad_sentences.csv', index=False)
corpora_v4_s.to_csv("data/extractions/new_data/corpora_v6.csv", index=False)